dlt/transformations/transform.py

```diff
@@ -0,0 +1,195 @@
+    table_format: TTableFormat = None,
+    references: TTableReferenceParam = None,
+    selected: bool = True,
+    spec: Type[TransformConfiguration] = None,
+    parallelized: bool = False,
+    incremental: Optional[TIncrementalConfig] = None,
+    standalone: bool = False,
+    # transformation specific
+    transformation_type: TTransformationType = None,
+    chunk_size: int = None,
+    lineage_mode: TLineageMode = None,
+) -> DltTransformResource:
+    # resolve defaults etc
+    chunk_size = chunk_size or DEFAULT_CHUNK_SIZE
+    lineage_mode = lineage_mode or "best_effort"
+
+    # resolve type of transformation
+    transformation_function_type = "python" if inspect.isgeneratorfunction(func) else "sql"
+    resolved_transformation_type = transformation_type
+    if not resolved_transformation_type:
+        resolved_transformation_type = "sql" if transformation_function_type == "sql" else "python"
+
+    # some sanity checks
+    if resolved_transformation_type == "sql" and transformation_function_type == "python":
+        raise TransformationTypeMismatch(
+            "Sql transformation must return single sql query string or dlt "
+            + "readablerelation and not be a generator function"
+        )
+
+    if lineage_mode in ["strict"] and transformation_function_type == "python":
+        raise LineageFailedException(
+            "Lineage mode 'strict' is only supported for sql transformations"
+        )
+
+    # build transform function
+    def transform_function(*args: Any, **kwargs: Any) -> Iterator[TDataItems]:
+        # collect all datasets from args
+        datasets: List[ReadableDBAPIDataset] = []
+        for arg in args:
+            if isinstance(arg, ReadableDBAPIDataset):
+                datasets.append(arg)
+
+        # TODO: implement is phyiscal destination check
+        # NOTE: there may be cases where some other dataset is used to get a starting
+        # point and it will be on a different destination.
+        # if datasets:
+        #     for d in datasets:
+        #         if not d.is_same_phyiscal_destination(datasets[0]):
+        #             raise IncompatibleDatasetsException(
+        #                 "All datasets used in transformation must be on the"
+        #                 + " same physical destination."
+        #             )
+
+        # Check that if sql transformation is used, destination dataset is on
+        # same physical destination
+        # if resolved_transformation_type == "sql" and not datasets[0].is_same_phyiscal_destination(
+        #     dlt.current.pipeline().dataset()
+        # ):
+        #     raise IncompatibleDatasetsException(
+        #         "When using sql transformations, the destination dataset"
+        #         + " must be on the same physical destination."
+        #     )
+
+        # extract query from transform function
+        select_query: str = None
+        transformation_result: Any = func(*args, **kwargs)
```

Comment:
we won't need most of the code if we remove the `transformation` arguments that I mentioned earlier